### ISSUE TITLE
Speed difference when use xpos.LowerTriangularMask() and attention_mask

### ISSUE THREAD

**Issue by Shannen3206** [facebookresearch/xformers#853]
# ❓ Questions and Help
When i use `xpos.memory_efficient_attention(q, k, v, attn_bias=xpos.LowerTriangularMask())`, The computation time of an attention procedure is `0.0002970695495605469`.
```python
start = time.time()
attn_output = xpos.memory_efficient_attention(q, k, v, attn_bias=xpos.LowerTriangularMask())
end = time.time()
print(end - start)
输出：0.0002970695495605469
```

However, this mask mechanism is different from the standard attention mask, causing the different inference results. Therefore, I take the following approach:
```python
attn_weights = None
query_states = query_states.transpose(1, 2)
key_states = key_states.transpose(1, 2)
value_states = value_states.transpose(1, 2)

add8 = 8 - attention_mask.shape[-1] % 8
attention_mask = attention_mask.expand(attention_mask.shape[0], self.num_heads, attention_mask.shape[2], attention_mask.shape[3])
tmp_attention_mask = torch.zeros([attention_mask.shape[0],
    attention_mask.shape[1],
    attention_mask.shape[2],
    attention_mask.shape[3]+add8]).type(query_states.type()).to(query_states.get_device())
tmp_attention_mask[:, :, :, :attention_mask.shape[-1]] = attention_mask

start  = time.time()
attn_output = xops.memory_efficient_attention(
    query_states, key_states, value_states, attn_bias=tmp_attention_mask[:, :, :, :attention_mask.shape[-1]]
)
end = time.time()
print(end - start)
输出：0.4382147789001465
```
Who can help solve this problem？

**Comment by danthe3rd** [comment]
Hi,
Your speed measurements are not done properly. You need to add `torch.cuda.synchronize()` before calling `time.time()`

> However, this mask mechanism is different from the standard attention mask

What do you mean? This is the standard causal attention mask used generally by LLMs. What mask do you need?

**Comment by Shannen3206** [comment]
Thank you for your advice on speed measurements. And I found the difference to be within tolerable limits. It was a mistake before. 

But I still have some questions.
- What's the difference between standard attention_mask and xops.LowerTriangularMask(). Why do the two methods produce large difference in the case of batch reasoning.
- Why float32 precision can produce NaN, and float16 can't.
- Last but not least: about the process of attention_mask in Method3.

I tried the following with the LlamaAttention function:
- Method 1: standard multi head attention;
- Method 2: flash attention with attn_bias=xops.LowerTriangularMask();
- Method 3: flash attention with attn_bias=tmp_attention_mask[:, :, :, :attention_mask.shape[-1];
- (Methods 4 and 5 are my other attempts, please ignore them)

```python
def forward(
    self,
    hidden_states: torch.Tensor,
    attention_mask: Optional[torch.Tensor] = None,
    position_ids: Optional[torch.LongTensor] = None,
    past_key_value: Optional[Tuple[torch.Tensor]] = None,
    output_attentions: bool = False,
    use_cache: bool = False,
) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
    bsz, q_len, _ = hidden_states.size()

    query_states = self.q_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
    key_states = self.k_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
    value_states = self.v_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)

    kv_seq_len = key_states.shape[-2]
    if past_key_value is not None:
        kv_seq_len += past_key_value[0].shape[-2]
    cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
    query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
    # [bsz, nh, t, hd]

    if past_key_value is not None:
        # reuse k, v, self_attention
        key_states = torch.cat([past_key_value[0], key_states], dim=2)
        value_states = torch.cat([past_key_value[1], value_states], dim=2)

    past_key_value = (key_states, value_states) if use_cache else None



    """方案1：标准注意力multi head attention"""
    torch.cuda.synchronize()
    start = time.time()
    attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)

    if attn_weights.size() != (bsz, self.num_heads, q_len, kv_seq_len):
        raise ValueError(
            f"Attention weights should be of size {(bsz * self.num_heads, q_len, kv_seq_len)}, but is"
            f" {attn_weights.size()}"
        )

    if attention_mask is not None:
        if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
            raise ValueError(
                f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
            )
        attn_weights = attn_weights + attention_mask
        attn_weights = torch.max(attn_weights, torch.tensor(torch.finfo(attn_weights.dtype).min))

    # upcast attention to fp32
    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
    attn_output = torch.matmul(attn_weights, value_states)

    if attn_output.size() != (bsz, self.num_heads, q_len, self.head_dim):
        raise ValueError(
            f"`attn_output` should be of size {(bsz, self.num_heads, q_len, self.head_dim)}, but is"
            f" {attn_output.size()}"
        )

    attn_output1 = attn_output.transpose(1, 2)
    torch.cuda.synchronize()
    end = time.time()
    print("方案1:\t", end - start)



    """方案2：flash attention 【attn_bias=xops.LowerTriangularMask()】"""
    attn_weights = None
    query_states_trans = query_states.transpose(1, 2)
    key_states_trans = key_states.transpose(1, 2)
    value_states_trans = value_states.transpose(1, 2)
    
    torch.cuda.synchronize()
    start = time.time()
    attn_output2 = xops.memory_efficient_attention(
        query_states_trans, key_states_trans, value_states_trans, attn_bias=xops.LowerTriangularMask(),
    )
    torch.cuda.synchronize()
    end = time.time()
    print("方案2:\t", end - start)



    """方案3：flash attention 【attn_bias = attention mask】"""
    attn_weights = None
    query_states_trans = query_states.transpose(1, 2)
    key_states_trans = key_states.transpose(1, 2)
    value_states_trans = value_states.transpose(1, 2)

    add8 = 8 - attention_mask.shape[-1] % 8
    tmp_attention_mask = attention_mask.expand(attention_mask.shape[0], self.num_heads, attention_mask.shape[2], attention_mask.shape[3])
    tmp_attention_mask = torch.zeros([tmp_attention_mask.shape[0],
                                        tmp_attention_mask.shape[1],
                                        tmp_attention_mask.shape[2],
                                        tmp_attention_mask.shape[3]+add8]).type(query_states.type()).to(query_states.get_device())
    tmp_attention_mask[:, :, :, :attention_mask.shape[-1]] = attention_mask

    torch.cuda.synchronize()
    start  = time.time()
    attn_output3 = xops.memory_efficient_attention(
        query_states_trans, key_states_trans, value_states_trans, attn_bias=tmp_attention_mask[:, :, :, :attention_mask.shape[-1]]
    )
    torch.cuda.synchronize()
    end = time.time()
    print("方案3:\t", end - start)



    """方案4：multi head attention 【no attention mask】"""
    torch.cuda.synchronize()
    start = time.time()
    attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)
    
    if attn_weights.size() != (bsz, self.num_heads, q_len, kv_seq_len):
        raise ValueError(
            f"Attention weights should be of size {(bsz * self.num_heads, q_len, kv_seq_len)}, but is"
            f" {attn_weights.size()}"
        )
    
    if attention_mask is not None:
        if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
            raise ValueError(
                f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
            )
        attn_weights = attn_weights + attention_mask
        attn_weights = torch.max(attn_weights, torch.tensor(torch.finfo(attn_weights.dtype).min))
    
    # upcast attention to fp32
    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query_states.dtype)
    attn_output = torch.matmul(attn_weights, value_states)
    
    # if attn_output.size() != (bsz, self.num_heads, q_len, self.head_dim):
    #     raise ValueError(
    #         f"`attn_output` should be of size {(bsz, self.num_heads, q_len, self.head_dim)}, but is"
    #         f" {attn_output.size()}"
    #     )
    
    attn_output4 = attn_output.transpose(1, 2)
    torch.cuda.synchronize()
    end = time.time()
    print("方案4:\t", end - start)
    


    """方案5：flash attention 【attn_bias=None】"""
    attn_weights = None
    query_states_trans = query_states.transpose(1, 2)
    key_states_trans = key_states.transpose(1, 2)
    value_states_trans = value_states.transpose(1, 2)
    
    torch.cuda.synchronize()
    start = time.time()
    attn_output5 = xops.memory_efficient_attention(
        query_states_trans, key_states_trans, value_states_trans, attn_bias=None,
    )
    torch.cuda.synchronize()
    end = time.time()
    print("方案5:\t", end - start)
    
    """误差计算"""
    print((attn_output1 - attn_output2).abs().mean())
    print((attn_output1 - attn_output3).abs().mean())
    print((attn_output4 - attn_output5).abs().mean())


    attn_output = attn_output3
    attn_output = attn_output.reshape(bsz, q_len, self.hidden_size)
    #
    attn_output = self.o_proj(attn_output)

    if not output_attentions:
        attn_weights = None

    return attn_output, attn_weights, past_key_value
```

① Why the difference between method 1 and method 2 is larger, and the difference between method1 and method 3 is smaller. What's the principle difference between method2 and method3

The following picture shows the situation of float16 batch reasoning. 
![image](https://github.com/facebookresearch/xformers/assets/53463361/62b77f39-a7e5-44ad-a505-2ddabda6f9a8)
The inference results of a batch of MMLU dataset is:
```
Method1: ["\nC\n\nThere", "\nA\n\nThere", "\nB\n\nThere", "\nC\n\nThere"]
Method2: ["2, and 2", "2 , ,", "\nB\n\nThere", "\nA\n\nThere"]
Method3: ["\nC\n\nThere", "\nA\n\nThere", "\nB\n\nThere", "\nC\n\nThere"]
```

② Why float32 precision can produce NaN, and float16 can't.

The following pictures show the situation of float32 batch reasoning. 
![image](https://github.com/facebookresearch/xformers/assets/53463361/8e4056d1-087a-440d-9e89-92c7cd718f4e)
![image](https://github.com/facebookresearch/xformers/assets/53463361/27df58bc-5f4f-49da-8f95-29f0627d451f)

③ How to process attention_mask quickly

The following operation is found to be much slower than the `xops.memory_efficient_attention operation` itself. Is there any quick way to do this?
```python
    add8 = 8 - attention_mask.shape[-1] % 8
    tmp_attention_mask = attention_mask.expand(attention_mask.shape[0], self.num_heads, attention_mask.shape[2], attention_mask.shape[3])
    tmp_attention_mask = torch.zeros([tmp_attention_mask.shape[0],
                                        tmp_attention_mask.shape[1],
                                        tmp_attention_mask.shape[2],
                                        tmp_attention_mask.shape[3]+add8]).type(query_states.type()).to(query_states.get_device())
    tmp_attention_mask[:, :, :, :attention_mask.shape[-1]] = attention_mask
```

**Comment by danthe3rd** [comment]
> * What's the difference between standard attention_mask and xops.LowerTriangularMask(). Why do the two methods produce large difference in the case of batch reasoning.
> What's the principle difference between method2 and method3

When you pass `LowerTriangularMask`, no bias *tensor* is created. It's passed as a flag to the kernel, and this means we have some memory reads, because we don't read a tensor from memory. So it should be faster.
In addition to that, we also know we dont have to compute the upper-right part of the attention matrix, so we can also skip half of the calculations.


> Why the difference between method 1 and method 2 is larger, and the difference between method1 and method 3 is smaller.

Method 2 and 3 compute the exact same thing. However, some small numerical differences are expected between different kernels/implementations, but shouldn't matter when training or doing inference end to end.

> * Why float32 precision can produce NaN, and float16 can't.

This shouldn't be the case and is a bug, unless there is an issue with your bias (contains nans, or entirely masks out a line).
If you can open an issue with a small repro we can have a look.
